utils/serde/src/main/java/io/apicurio/registry/utils/serde/AbstractKafkaSerializer.java

```diff
@@ -65,12 +74,18 @@ public AbstractKafkaSerializer(
             String artifactId = getArtifactIdStrategy().artifactId(topic, isKey(), schema);
             long id = getGlobalIdStrategy().findId(getClient(), artifactId, artifactType(), schema);
             ByteArrayOutputStream out = new ByteArrayOutputStream();
-            out.write(MAGIC_BYTE);
-            getIdHandler().writeId(id, out);
-            serializeData(schema, data, out);
+            if (headerUtils != null) {
+                headerUtils.addSchemaHeaders(headers, artifactId, id);
+                serializeData(headers, schema, data, out);
+            } else {
+                out.write(MAGIC_BYTE);
+                getIdHandler().writeId(id, out);
+                serializeData(schema, data, out);
+            }
```

Comment:
I was expecting the decision point to be a configuration property, not the presence of the `Headers` object - what determines which variant of `serialize` is called?

Reply:
The headerUtil is effectively only created if the config has been set